### PR TITLE
simplify skip-if tests >> 0/n

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

simplify chaining skip-if in pytests

```python
@pytest.mark.skipif(**skipif_args(min_gpus=2, min_torch="99"))
def test_skip():
    ...
```    

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
Why not make it `@skipIfNoGPUs` instead of `@pytest.mark.skipif(**_SKIPIF_NO_GPUS)`?

**Comment by Borda** [comment]
> Why not make it `@skipIfNoGPUs` instead of `@pytest.mark.skipif(**_SKIPIF_NO_GPUS)`?

the main point was about chaining them together...

File: tests/accelerators/legacy/test_ddp.py [hunk 1]
```diff
@@ -96,7 +97,7 @@ def test_cli(tmpdir):
 # END: test_cli ddp test
 
 
-@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.skipif(**_SKIPIF_NO_GPUS)
```
Comment (awaelchli) [inline tests/accelerators/legacy/test_ddp.py:100]: it's not equivalent to saying <2 gpus

Comment (Borda) [reply tests/accelerators/legacy/test_ddp.py:100]: it is, note difference `_skipif_no_gou` and
 `_skipif_no_gouS`

**Comment by carmocca** [comment]
> the main point was about chaining them together and on the same line...

IMO I'd rather have different conditions in different lines, as is usual in other projects

File: tests/accelerators/legacy/test_ddp.py (hunk 1 above)
Comment (awaelchli) [reply tests/accelerators/legacy/test_ddp.py:100]: This is a bit of a problem with natural language here. 
May I suggest a decorator with argument? For example:
`@device_required(gpus=2)`
or
`@gpu_required(2)`
or something like that?

**Comment by Borda** [comment]
> > the main point was about chaining them together and on the same line...
> 
> IMO I'd rather have different conditions in different lines, as is usual in other projects

so to have grouped under unitest class, since we have for an example test file where each of the included tests has 5 skips if and all the same for all tests so in the skip is most of the code and not clear reading...
cc: @awaelchli @SeanNaren

File: tests/accelerators/legacy/test_ddp.py (hunk 1 above)
Comment (Borda) [reply tests/accelerators/legacy/test_ddp.py:100]: well yes, my main motivation was king of simple chaining conditions to be easier to read it...

Comment (Borda) [reply tests/accelerators/legacy/test_ddp.py:100]: well, @awaelchli but for your two you in fact mean `min_gpus=2`

Comment (awaelchli) [reply tests/accelerators/legacy/test_ddp.py:100]: yes, that should be fine. If 2 are required we are also happy with >=2. We could name the positional arg min_gpus

**Comment by Borda** [comment]
Ok, so there are let's say three ways to go:
1. just to standardize the skip if content so we avoid typos and misalignment of the condition and skip reason so the usage would be just 
    ```
    @pytest.mark.skipif(**_SKIPIF_PT_LE_1_6)
    @pytest.mark.skipif(**_SKIPIF_NO_APEX)
    @pytest.mark.skipif(**_SKIPIF_NO_GPUS)
    def test_...(...):
        ...
    ```
2. chaining the skip conditions so it would be a single line and easier to read
    ```
    @pytest.mark.skipif(**(_SKIPIF_PT_LE_1_6 | _SKIPIF_NO_APEX | _SKIPIF_NO_GPUS))
    def test_...(...):
        ...
    ```
3. creating custom wrapper with all these cases as argument (which would increase contribution level as a user needs to read/learn this custom parameterization)
    ```
    @lightning_mark_skipif(min_pt=1.6, apex=True, min_gpus=2)
    def test_...(...):
        ...
    ```

File: tests/accelerators/legacy/test_ddp.py (hunk 1 above)
Comment (Borda) [reply tests/accelerators/legacy/test_ddp.py:100]: mind comment on https://github.com/PyTorchLightning/pytorch-lightning/pull/5920#issuecomment-781435819

**Comment by Borda** [comment]
@tadejsv @SeanNaren @SkafteNicki @akihironitta @rohitgr7 it is not exactly like the promise above as it turned out that the overwriting pytest marks would be tricky and no very sure of any accident unwanted interaction with other pytest marks such as parameterized...

**Comment by Borda** [comment]
after merging this one I will apply this simplification to the remaining codebase...

File: tests/helpers/skipif.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        torch_version = LooseVersion(get_distribution("torch").version)
+        conditions.append(torch_version < LooseVersion(min_torch))
+        reasons.append(f"minimal version `torch>={min_torch}")
+
+    if pt_quant:
+        conditions.append(not _TORCH_QUANTIZE_AVAILABLE or _MISS_QUANT_DEFAULT)
+        reasons.append("PyTorch quantization")
+
+    if not any(conditions):
+        return dict(condition=False, reason="no reason, just go test it...")
```
Comment (kaushikb11) [inline tests/helpers/skipif.py:60]:
```suggestion
        return dict(condition=False, reason="Conditions satisfied, going ahead with the test")
```

File: tests/helpers/skipif.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    >>> pprint(create_skipif(min_torch="99"))
+    {'condition': True, 'reason': 'test requires minimal version `torch>=99'}
+    >>> pprint(create_skipif(min_torch="0.0"))
+    {'condition': False, 'reason': 'no reason, just go test it...'}
+    """
+    conditions = []
+    reasons = []
+
+    if min_gpus:
+        conditions.append(torch.cuda.device_count() < 2)
```
Comment (kaushikb11) [inline tests/helpers/skipif.py:47]:
```suggestion
        conditions.append(torch.cuda.device_count() < min_gpus)
```

**Review by kaushikb11 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -224,29 +223,23 @@ def predict(self, args: List[Union[Any, int]]) -> _STEP_OUTPUT_TYPE:
         with self.precision_plugin.predict_context(), self.training_type_plugin.predict_context():
             return self.training_type_plugin.predict(*args)
 
-    def training_step_end(
-        self, output: _STEP_OUTPUT_TYPE
-    ) -> _STEP_OUTPUT_TYPE:
+    def training_step_end(self, output: _STEP_OUTPUT_TYPE) -> _STEP_OUTPUT_TYPE:
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:226]: Are those formatting changes really necessary in this PR? they shouldn't be touched at all?

File: pytorch_lightning/trainer/callback_hook.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 from abc import ABC
 from copy import deepcopy
 from inspect import signature
-from typing import List, Dict, Any, Type, Callable
+from typing import Any, Callable, Dict, List, Type
```
Comment (justusschock) [inline pytorch_lightning/trainer/callback_hook.py:18]: Also just formatting

File: tests/helpers/skipif.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    if min_gpus:
+        conditions.append(torch.cuda.device_count() < min_gpus)
+        reasons.append(f"multi-GPU machine with at least {min_gpus}")
+
+    if min_torch:
+        torch_version = LooseVersion(get_distribution("torch").version)
+        conditions.append(torch_version < LooseVersion(min_torch))
+        reasons.append(f"minimal version `torch>={min_torch}")
+
+    if pt_quant:
```
Comment (justusschock) [inline tests/helpers/skipif.py:55]: do we also want to add stuff like this for fairscale, horovod, deepspeed etc.?

also isn't this basically also a check for a pytorch version here?

**Review by justusschock — APPROVED** [review]

File: tests/helpers/skipif.py (hunk 6 above)
Comment (Borda) [reply tests/helpers/skipif.py:55]: Yes I will add also all others in follow up PRs...
This is for the concept

File: tests/helpers/skipif.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        @pytest.mark.skipif(**create_skipif(min_torch="99"))
+        def test_any_func(...):
+            ...
+
+    >>> from pprint import pprint
+    >>> pprint(create_skipif(min_torch="99"))
+    {'condition': True, 'reason': 'test requires minimal version `torch>=99'}
+    >>> pprint(create_skipif(min_torch="0.0"))
+    {'condition': False, 'reason': 'no reason, just go test it...'}
```
Comment (Borda) [inline tests/helpers/skipif.py:41]:
```suggestion
    {'condition': False, 'reason': 'Conditions satisfied, going ahead with the test.'}
```

File: tests/helpers/skipif.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        torch_version = LooseVersion(get_distribution("torch").version)
+        conditions.append(torch_version < LooseVersion(min_torch))
+        reasons.append(f"minimal version `torch>={min_torch}")
+
+    if pt_quant:
+        conditions.append(not _TORCH_QUANTIZE_AVAILABLE or _MISS_QUANT_DEFAULT)
+        reasons.append("PyTorch quantization")
+
+    if not any(conditions):
+        return dict(condition=False, reason="Conditions satisfied, going ahead with the test")
```
Comment (Borda) [inline tests/helpers/skipif.py:60]:
```suggestion
        return dict(condition=False, reason="Conditions satisfied, going ahead with the test.")
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/accelerators/accelerator.py:226]: Well, since will drop formating checks from particular PRs, I don't have many other options how to keep codebase healthy

File: tests/helpers/skipif.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        @pytest.mark.skipif(**create_skipif(min_torch="99"))
+        def test_any_func(...):
+            ...
+
+    >>> from pprint import pprint
+    >>> pprint(create_skipif(min_torch="99"))
+    {'condition': True, 'reason': 'test requires minimal version `torch>=99'}
+    >>> pprint(create_skipif(min_torch="0.0"))
+    {'condition': False, 'reason': 'Conditions satisfied, going ahead with the test.'}
```
Comment (SkafteNicki) [inline tests/helpers/skipif.py:41]: doctest is failing here, should probably be:
```suggestion
    {'condition': False, 'reason': # doctest: +NORMALIZE_WHITESPACE
     'Conditions satisfied, going ahead with the test.'}
```

File: tests/helpers/skipif.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    min_torch: Optional[str] = None,
+    pt_quant: bool = False,
+) -> dict:
+    """ Creating aggregated arguments for standatd pytest skipif, sot the usecase is::
+
+        @pytest.mark.skipif(**create_skipif(min_torch="99"))
+        def test_any_func(...):
+            ...
+
+    >>> from pprint import pprint
```
Comment (SkafteNicki) [inline tests/helpers/skipif.py:37]: could you also provide an example that shows the output with two conditions

**Review by SkafteNicki — APPROVED** [review]
LGTM :]

File: tests/helpers/skipif.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import pytest
+import torch
+from pkg_resources import get_distribution
+
+from pytorch_lightning.utilities import _TORCH_QUANTIZE_AVAILABLE
+
+_MISS_QUANT_DEFAULT = 'fbgemm' not in torch.backends.quantized.supported_engines
+
+
+def create_skipif(
```
Comment (carmocca) [inline tests/helpers/skipif.py:26]: maybe rename to `conditions` or `skipif_args` as this doesn't really create the `skipif`

File: tests/helpers/skipif.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+from distutils.version import LooseVersion
+from typing import Optional
+
+import pytest
+import torch
+from pkg_resources import get_distribution
+
+from pytorch_lightning.utilities import _TORCH_QUANTIZE_AVAILABLE
+
+_MISS_QUANT_DEFAULT = 'fbgemm' not in torch.backends.quantized.supported_engines
```
Comment (carmocca) [inline tests/helpers/skipif.py:23]: put this inside the `if pt_quant` block?

File: tests/helpers/skipif.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def create_skipif(
+    min_gpus: int = 0,
+    min_torch: Optional[str] = None,
+    pt_quant: bool = False,
```
Comment (carmocca) [inline tests/helpers/skipif.py:29]: call this `quantization` or `quantization_available`?

File: tests/helpers/skipif.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+    if pt_quant:
+        conditions.append(not _TORCH_QUANTIZE_AVAILABLE or _MISS_QUANT_DEFAULT)
+        reasons.append("PyTorch quantization")
+
+    if not any(conditions):
+        return dict(condition=False, reason="Conditions satisfied, going ahead with the test.")
+
+    reasons = [rs for cond, rs in zip(conditions, reasons) if cond]
+    reason = "test requires " + ' + '.join(reasons)
```
Comment (carmocca) [inline tests/helpers/skipif.py:63]: Can we have a simpler and shorter reason? something like:

`Required: [gpus>2, torch>=1.7.0, quantization=available]`

instead of the current: 

`test requires multi-GPU machine with at least 2 + minimal version torch>=1.7.0 + PyTorch quantization

**Review by carmocca — COMMENTED** [review]
So why don't we provide the skipif ourselves?

`return pytest.mark.skipif(condition=any(conditions), reason=reason)`

does it not work?

File: tests/helpers/skipif.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+_MISS_QUANT_DEFAULT = 'fbgemm' not in torch.backends.quantized.supported_engines
+
+
+def create_skipif(
+    min_gpus: int = 0,
+    min_torch: Optional[str] = None,
+    pt_quant: bool = False,
+) -> dict:
+    """ Creating aggregated arguments for standatd pytest skipif, sot the usecase is::
```
Comment (akihironitta) [inline tests/helpers/skipif.py:31]:
```suggestion
    """ Creating aggregated arguments for standard pytest skipif, sot the usecase is::
```

**Review by akihironitta — APPROVED** [review]

File: tests/helpers/skipif.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+        @pytest.mark.skipif(**create_skipif(min_torch="99"))
+        def test_any_func(...):
+            ...
+
+    >>> from pprint import pprint
+    >>> pprint(skipif_args(min_torch="99"))
+    {'condition': True, 'reason': 'Required: [torch>=99]'}
+    >>> pprint(skipif_args(min_torch="0.0"))
```
Comment (Borda) [inline tests/helpers/skipif.py:38]:
```suggestion
```

File: tests/helpers/skipif.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+import torch
+from pkg_resources import get_distribution
+
+from pytorch_lightning.utilities import _TORCH_QUANTIZE_AVAILABLE
+
+
+def skipif_args(
+    min_gpus: int = 0,
+    min_torch: Optional[str] = None,
+    avail_quant: bool = False,
```
Comment (carmocca) [inline tests/helpers/skipif.py:27]:
```suggestion
    quant_available: bool = False,
```

maybe not shorten `avail`? I don't think it's intuitive that `avail == available` for non-english speakers

File: tests/helpers/skipif.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+    if min_torch:
+        torch_version = LooseVersion(get_distribution("torch").version)
+        conditions.append(torch_version < LooseVersion(min_torch))
+        reasons.append(f"torch>={min_torch}")
+
+    if avail_quant:
+        _miss_default = 'fbgemm' not in torch.backends.quantized.supported_engines
+        conditions.append(not _TORCH_QUANTIZE_AVAILABLE or _miss_default)
+        reasons.append("PyTorch quantization is available")
```
Comment (carmocca) [inline tests/helpers/skipif.py:56]:
```suggestion
        reasons.append("pytorch_quantization=available")
```

Comment (Borda) [reply tests/helpers/skipif.py:56]: that is not clear what it means...

File: tests/helpers/skipif.py (hunk 17 above)
Comment (Borda) [reply tests/helpers/skipif.py]: done

**Review by SeanNaren — APPROVED** [review]
Nice. Should clean up some of these really stacked decorators finally, so nice work!
